### PR TITLE
Multi-head 1:N + Multi-head in avro-style

### LINKED ISSUE #105: Multi-head connections with type "1:N" is not supported
Currently qraphql schema with multi-head connections with type "1:N" is not valid.

### LINKED ISSUE #118: Use avro-style union for multi-head connections
It allows to represent query shape using avro-schema and will be consistent with representation of union fields.

Avro-style union means single-field object whose field name is type determinant.

### PR DESCRIPTION
**Pull request by SudoBobo**
Closes #105, closes #118

### CODE REVIEW

File: graphql/tarantool_graphql.lua [hunk 1]
```diff
@@ -449,26 +493,20 @@ end
 --- connection
 local function convert_simple_connection(state, connection, collection_name)
```
Comment (SudoBobo) [inline graphql/tarantool_graphql.lua:495]: All changes in this function is refactoring changes; Tried to make it a bit easier to read

File: graphql/tarantool_graphql.lua [hunk 2]
```diff
@@ -612,21 +722,22 @@ local function convert_union_connection(state, connection, collection_name)
             'Determinant keys:\n"%s"'):
             format(yaml.encode(parent), yaml.encode(determinant_keys)))
 
-        local variant_num
-        local resulting_variant
-        for i, variant in ipairs(c.variants) do
-            variant_num = i
-            local is_match = utils.is_subtable(parent, variant.determinant)
-
+        local var_idx
```
Comment (SudoBobo) [inline graphql/tarantool_graphql.lua:726]: Just changed variables names

File: graphql/tarantool_graphql.lua [hunk 3]
```diff
@@ -369,6 +369,50 @@ local function specify_destination_type(destination_type, connection_type)
...
+--- Why the 'boxing' of collection types is needed and how it is done is
+--- described in comments to @{convert_multihead_connection}.
+---
+--- @tparam table type_to_box GraphQL Object type (which represents a collection)
+--- @tparam string connection_type of given collection (1:1, 1:1* or 1:N)
+--- @tparam string type_to_box_name name of given 'type_to_box' (It can not
+--- be taken from 'type_to_box' because at the time of function execution
+--- 'type_to_box' refers to an empty table, which later will be filled with
+--- actual type table)
+--- @treturn table GraphQL Object type represent 'boxed' collection
```
Comment (Totktonada) [inline graphql/tarantool_graphql.lua]: Two values are returned from the function.

File: graphql/tarantool_graphql.lua [hunk 4]
```diff
@@ -559,45 +601,113 @@ local function convert_simple_connection(state, connection, collection_name)
...
     for _, v in ipairs(c.variants) do
         assert(v.determinant, 'each variant should have a determinant')
-            assert(type(v.determinant) == 'table', 'variant\'s determinant ' ..
-            'must end be a table, got ' .. type(v.determinant))
-        assert(type(v.destination_collection) == 'string',
-            'variant.destination_collection must be a string, got ' ..
-            type(v.destination_collection))
-        assert(type(v.parts) == 'table',
-            'variant.parts must be a table, got ' .. type(v.parts))
+        check (v.determinant, 'variant\'s determinant', 'table')
```
Comment (Totktonada) [inline graphql/tarantool_graphql.lua]: Redundant whitespace before open parenthesis.

File: graphql/tarantool_graphql.lua [hunk 5]
```diff
@@ -559,45 +601,113 @@ local function convert_simple_connection(state, connection, collection_name)
...
+--- In case of 1:N {...} indicates a list of instances. Using of 'boxing'
+--- provides the needed format.
 ---
 --- @tparam table state for collection types
---- @tparam table connection union connection to create field on
+--- @tparam table connection multi-head connection to create GraphQL Union on
 --- @tparam table collection_name name of the collection which has given
 --- connection
-local function convert_union_connection(state, connection, collection_name)
+--- @treturn table GraphQL Union type
```
Comment (Totktonada) [inline graphql/tarantool_graphql.lua:681]: Three results are returned from the function.

File: graphql/tarantool_graphql.lua [hunk 6]
```diff
@@ -720,33 +839,33 @@ local convert_connection_to_field = function(state, connection, collection_name)
...
---- was used to create `gql_type`. `avro_name` is used to provide avro-valid names
---- for fields of boxed types
+--- @tparam table type_to_box GraphQL type to be boxed
+--- @tparam string box_field_name name of the single box field
 --- @treturn table GraphQL Object
-local function box_type(gql_type, avro_name)
-    check(gql_type, 'gql_type', 'table')
+local function box_type(type_to_box, box_field_name)
+    check(type_to_box, 'type_to_box', 'table')
+    check(box_field_name, 'type_to_box_name', 'string')
```
Comment (Totktonada) [inline graphql/tarantool_graphql.lua]: So, box_field_name or type_to_box_name?

File: test/testdata/multihead_conn_testdata.lua [hunk 7]
```diff
@@ -149,47 +245,115 @@ function multihead_conn_testdata.init_spaces()
...
     box.space.starship_collection:drop()
     box.space.hero_collection:drop()
+    box.space.credit_account_collection:drop()
+    box.space.dublon_account_collection:drop()
 end
 
 function multihead_conn_testdata.run_queries(gql_wrapper)
     local test = tap.test('multihead_conn')
-    test:plan(2)
+    test:plan(1)
```
Comment (Totktonada) [inline test/testdata/multihead_conn_testdata.lua]: It seems that two tests are runned here.

**Review by Totktonada — APPROVED** [review]
Looks good. Please, fix minor comments below and proceed with the patch.

In the addition to the comments below, the following notes I have to say:

* Please, rebase on master and make sure CI is okay.
* Please, squash the commits and fix `105` → `#105` in the commit message.

File: graphql/tarantool_graphql.lua (hunk 3 above)
Comment (SudoBobo) [reply graphql/tarantool_graphql.lua]: Fxd.

File: graphql/tarantool_graphql.lua (hunk 5 above)
Comment (SudoBobo) [reply graphql/tarantool_graphql.lua:681]: Are you sure? Three result are returned from the `resolve_variant`
`convert_union_connection` returns only one.

File: graphql/tarantool_graphql.lua (hunk 4 above)
Comment (SudoBobo) [reply graphql/tarantool_graphql.lua]: Fxd.

File: test/testdata/multihead_conn_testdata.lua (hunk 7 above)
Comment (SudoBobo) [reply test/testdata/multihead_conn_testdata.lua]: Fxd.

File: graphql/tarantool_graphql.lua (hunk 6 above)
Comment (SudoBobo) [reply graphql/tarantool_graphql.lua]: box_field_name. Fxd.

**Pull request merged**